Grid overlay

lbartoletti opened the pull request:
Rewrite GridLine and GridPolygon; Add an overlay option; add full tests; rm old tests.

This pull request close #3546 and #3547.

lbartoletti committed:
Rewrite GridLine and GridPolygon; Add an overlay option; add full tests; rm old tests

lbartoletti committed:
Merge branch 'master' of http://github.com/qgis/QGIS into grid_overlay

lbartoletti committed:
Fix vOverlay for GridLine; fix tests

lbartoletti committed:
readd test files

nyalldawson commented on python/plugins/processing/algs/qgis/GridLine.py:
```diff
@@ -116,52 +128,105 @@ def processAlgorithm(self, progress):
+            polyline.append(QgsPoint(x1, y))
+            polyline.append(QgsPoint(x2, y))
+
+            feat.setGeometry(QgsGeometry.fromPolyline(polyline))
```
fromPolyline is slow. Can you restore the implementation which uses QgsPointV2/QgsLineString?

lbartoletti replied:
I have restored the implementation that uses QgsPointV2 / QgsLineString for GridLine, I'll pull it tomorrow, this will depend on your answer to your second comment. Do you have the same suggestion for GridPolygon?

nyalldawson commented on python/plugins/processing/tests/testdata/expected/grid_diamond.gml:
```diff
@@ -1,214 +0,0 @@
-<?xml version="1.0" encoding="utf-8" ?>
```
It makes me nervous seeing the existing tests removed. Is there any reason these could not be retained? If not, have you checked to see what the differences in the resultant layers are?

lbartoletti replied:
You're right, I have not quite explained my proposal. 
There are some differences between the proposed alg and the original alg. Obviously this is a new option for overlay, but also:
* The point of origin is located at the bottom left and not at the top left as on the original algo; It may be an error.
* The original algo does not fully cover the extent for diamond/rectangle and does not necessarily end the grid for the lines, as user it is a point that I can regret.

That is why I proposed new tests.

However, I have prepared another version more current friendly where the existing tests are respected except for GridLine where I propose to totally grid the extent like GridPolygon.

lbartoletti committed:
I have rewrite algs from current. Just add an overlay option and not
change current algorithm. Actual tests are respected and only add one
test for each algs with overlay option.

lbartoletti committed:
add correct grid_hexagon_overlay test file

lbartoletti commented:
Travis says there is an error but, not for my test...
`        Start   4: ProcessingQgisAlgorithmsTest
  4/280 Test   #4: ProcessingQgisAlgorithmsTest ............   Passed    7.23 sec`

But at the end
`The job exceeded the maximum time limit for jobs, and has been terminated.`

m-kuhn commented:
@lbartoletti that was an unrelated issue with our testing infrastructure, your tests are all fine

lbartoletti committed:
Resolve conflicts from ede452b (Port feedback to c++). Just change
progress to feedback.

lbartoletti committed:
change setProgress to setPercentage

lbartoletti committed:
Merge branch 'master' into grid_overlay

nyalldawson commented:
Thanks @lbartoletti !
